File: TrappingRainWater.cpp
#include<vector>
#include<iostream>
using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        // int sum=0,max=0,start=0,end=0;
        // for(int i=0;i<height.size();i++){
        //     if(max<height[i]){
        //         sum+=max*(end-start);
        //         max=height[i];
        //         start=end=i;
        //     }else{
        //         end++;
        //     }
        // }   
        int sum=0;
        for(int start=0,end=height.size()-1;start<end;){
            if(height[start]>0&&height[end]>0){
                for(int i=start;i<=end;i++){
                    height[i]--;
                }
            }
            if(height[start]<=0)start++;
            if(height[end]<=0)end--;  
        }
        
        for(int i=0;i<height.size();i++){
            if(height[i]<0){
                sum+=-height[i];
            }
        }
        return sum;     
    }
};

File: Trapping_Rain_Water.kt
import java.util.*

class Solution {
    fun trap(height: IntArray): Int {
        var ans = 0
        val st = Stack<Int>()
        for ((i, h) in height.withIndex()) {
            while (!st.empty() && h > height[st.peek()]) {
                val middle = st.pop()
                if (st.empty()) break
                ans += (Math.min(h, height[st.peek()]) - height[middle]) * (i - st.peek() - 1)
            }
            st.push(i)
        }
        return ans
    }
}

fun main() {
    val solution = Solution()
    val testCases = arrayOf(intArrayOf(0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1))
    for (case in testCases) {
        println(solution.trap(case))
    }
}